[Docs][KubeRay] Update doc for kubectl-plugin (#49890)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/kubectl-plugin.md

(kubectl-plugin)=

> # Use kubectl Plugin (beta)

Reviewer: Can the casing here be consistent, either `Use kubectl plugin (beta)` or `Use Kubectl Plugin (beta)`

Author: Done!

## Example

> Most of these example assumes you have a Ray cluster running on Kubernetes. See {ref}`RayCluster Quickstart <kuberay-raycluster-quickstart>` if you don't have a Ray cluster running on Kubernetes.

Author: Done!

## Example

Most of these example assumes you have a Ray cluster running on Kubernetes. See {ref}`RayCluster Quickstart <kuberay-raycluster-quickstart>` if you don't have a Ray cluster running on Kubernetes.

### Get all Ray clusters

> ```shell

Reviewer 2: Maybe simply use plain text <code>```</code> if the code block containing logs because those logs do not need to be colored. Ditto for all other code blocks.

![image](https://github.com/user-attachments/assets/88885875-9a4b-4f22-b86d-19f65b8cf7ce)

Author: I changed all of the blocks to text except the code ones. PTAL

> & kubectl ray get cluster

Reviewer 2: ```suggestion
$ kubectl ray get cluster
```

Author: Done!

### Get Ray cluster logs

```text
$ kubectl ray logs sample-raycluster-kfhl6
No output directory specified, creating dir under current directory using cluster name.
Command set to retrieve both head and worker node logs.
Downloading log for Ray Node sample-raycluster-kfhl6-head-87xpb
Downloading log for Ray Node sample-raycluster-kfhl6-small-group-worker-54qfm
```

#### Get Ray job logs

> Users can also Ray logs with RayJobs and RayServices

Author: Done!

```text
$ kubectl ray logs rayjob/rayjob-interactivemode
No output directory specified, creating dir under current directory using resource name.
Command set to retrieve both head and worker node logs.
Downloading log for Ray Node rayjob-interactivemode-raycluster-qbkr8-head-dwm84
Downloading log for Ray Node rayjob-interactivemode-raycluster-qbkr8-small-grou-worker-hr2jp
```

> ### Create a Ray Cluster

Reviewer 3: ```suggestion
### Create a Ray cluster
```

Author: Done!

> This allows users to creates a valid RayCluster. The default values are follows:

Author: I updated the wording, PTAL

| Parameter | Default |
| -------- | ---------- |
| ray version | 2.39.0 |
| ray image | rayproject/ray:\<ray version\> |
| head cpu | 2 |
| head memory | 4Gi |
| worker replicas | 1 |
| worker cpu | 2 |
| worker memory | 4Gi |
| worker gpu | 0 |

Currently only one worker group is created.

```text
$ kubectl ray create cluster raycluster-sample
Created Ray Cluster: raycluster-sample
$ kubectl ray get cluster 
NAME                             NAMESPACE   DESIRED WORKERS   AVAILABLE WORKERS   CPUS   GPUS   TPUS   MEMORY   AGE
raycluster-sample                default     1                 1                   4      0      0      8Gi      25s
```

> ### Submit A Ray Job

Author: Done!

This is a wrapper around the `ray job submit` command. The plugin can automatically forward the ports to the Ray cluster and submit the job. This command can also provision a ephemeral cluster needed to execute the job if no RayJob is provided.

Assume that under the current directory, you have a file named `sample_code.py`.

```python
import ray
ray.init(address="auto")

@ray.remote
def f(x):
    return x * x

futures = [f.remote(i) for i in range(4)]
print(ray.get(futures)) # [0, 1, 4, 9]
```

> #### Submit Ray Job without yaml

Reviewer 3: ```suggestion
#### Submit a Ray job without a YAML file
```

Reviewer 3: Is this supposed to be a `Ray job` or a `RayJob`?

Author: I think `Ray job` is fine

> Users can submit a RayJob without specifying a yaml, the command generates a RayJob based on the following:

Reviewer 3: ```suggestion
You can submit a RayJob without specifying a YAML file. The command generates a RayJob based on the following:
```

Author: Done!

2025-01-06 11:53:33,631	INFO cli.py:292 -- ray job logs raysubmit_9NfCvwcmcyMNFCvX
2025-01-06 11:53:33,631	INFO cli.py:294 -- Query the status of the job:
2025-01-06 11:53:33,631	INFO cli.py:296 -- ray job status raysubmit_9NfCvwcmcyMNFCvX
2025-01-06 11:53:33,631	INFO cli.py:298 -- Request the job to be stopped:
2025-01-06 11:53:33,631	INFO cli.py:300 -- ray job stop raysubmit_9NfCvwcmcyMNFCvX
2025-01-06 11:53:33,786	INFO cli.py:307 -- Tailing logs until the job exits (disable with --no-wait):
2025-01-06 11:53:33,410	INFO job_manager.py:530 -- Runtime env is setting up.
2025-01-06 11:53:34,806	INFO worker.py:1494 -- Using address 10.12.0.9:6379 set in the environment variable RAY_ADDRESS
2025-01-06 11:53:34,806	INFO worker.py:1634 -- Connecting to existing Ray cluster at address: 10.12.0.9:6379...
2025-01-06 11:53:34,814	INFO worker.py:1810 -- Connected to Ray cluster. View the dashboard at 10.12.0.9:8265 
[0, 1, 4, 9]
2025-01-06 11:53:38,368	SUCC cli.py:63 -- ------------------------------------------
2025-01-06 11:53:38,368	SUCC cli.py:64 -- Job 'raysubmit_9NfCvwcmcyMNFCvX' succeeded
2025-01-06 11:53:38,368	SUCC cli.py:65 -- ------------------------------------------
```

> #### Submit a Ray Job with a specified RayJob yaml

Reviewer 3: ```suggestion
#### Submit a Ray job with a RayJob YAML
```

Author: Done!

> Users can also designate a specific RayJob yaml

Reviewer 3: ```suggestion
You can designate a specific RayJob YAML to submit a Ray job.
```

Author: Done!

> Users can add the following yaml file `ray-job.interactivemode.yaml`

Reviewer 3: ```suggestion
Add the following YAML file `ray-job.interactivemode.yaml`:
```

Author: Done!

2025-01-06 12:44:42,313	INFO cli.py:292 -- ray job logs raysubmit_fuBdjGnecFggejhR
2025-01-06 12:44:42,313	INFO cli.py:294 -- Query the status of the job:
2025-01-06 12:44:42,313	INFO cli.py:296 -- ray job status raysubmit_fuBdjGnecFggejhR
2025-01-06 12:44:42,313	INFO cli.py:298 -- Request the job to be stopped:
2025-01-06 12:44:42,313	INFO cli.py:300 -- ray job stop raysubmit_fuBdjGnecFggejhR
2025-01-06 12:44:42,472	INFO cli.py:307 -- Tailing logs until the job exits (disable with --no-wait):
2025-01-06 12:44:41,931	INFO job_manager.py:530 -- Runtime env is setting up.
2025-01-06 12:44:43,542	INFO worker.py:1494 -- Using address 10.12.0.10:6379 set in the environment variable RAY_ADDRESS
2025-01-06 12:44:43,542	INFO worker.py:1634 -- Connecting to existing Ray cluster at address: 10.12.0.10:6379...
2025-01-06 12:44:43,551	INFO worker.py:1810 -- Connected to Ray cluster. View the dashboard at 10.12.0.10:8265 
[0, 1, 4, 9]
2025-01-06 12:44:47,830	SUCC cli.py:63 -- ------------------------------------------
2025-01-06 12:44:47,830	SUCC cli.py:64 -- Job 'raysubmit_fuBdjGnecFggejhR' succeeded
2025-01-06 12:44:47,830	SUCC cli.py:65 -- ------------------------------------------
```

> ### Delete a Ray Cluster

Reviewer 3: ```suggestion
### Delete a Ray cluster
```

Author: Done!

### Get Ray cluster logs

```shell
$ kubectl ray logs sample-raycluster-kfhl6
No output directory specified, creating dir under current directory using cluster name.
Command set to retrieve both head and worker node logs.
Downloading log for Ray Node sample-raycluster-kfhl6-head-87xpb
Downloading log for Ray Node sample-raycluster-kfhl6-small-group-worker-54qfm
```

#### Get Ray job logs

Users can also Ray logs with RayJobs and RayServices

```shell
$ kubectl ray logs rayjob/rayjob-interactivemode
No output directory specified, creating dir under current directory using resource name.
Command set to retrieve both head and worker node logs.
Downloading log for Ray Node rayjob-interactivemode-raycluster-qbkr8-head-dwm84
Downloading log for Ray Node rayjob-interactivemode-raycluster-qbkr8-small-grou-worker-hr2jp
```

### Submit A Ray Job

> This is a wrapper around the `ray job submit` command. The plugin can automatically forward the ports to the Ray cluster and submit the job.

Reviewer: We should add more details here that running `kubectl ray job submit` not only runs `ray job submit`, but also provisions the ephemeral cluster used to execute the job

Author: Changed, PTAL!

2025-01-06 12:44:42,313	INFO cli.py:292 -- ray job logs raysubmit_fuBdjGnecFggejhR
2025-01-06 12:44:42,313	INFO cli.py:294 -- Query the status of the job:
2025-01-06 12:44:42,313	INFO cli.py:296 -- ray job status raysubmit_fuBdjGnecFggejhR
2025-01-06 12:44:42,313	INFO cli.py:298 -- Request the job to be stopped:
2025-01-06 12:44:42,313	INFO cli.py:300 -- ray job stop raysubmit_fuBdjGnecFggejhR
2025-01-06 12:44:42,472	INFO cli.py:307 -- Tailing logs until the job exits (disable with --no-wait):
2025-01-06 12:44:41,931	INFO job_manager.py:530 -- Runtime env is setting up.
2025-01-06 12:44:43,542	INFO worker.py:1494 -- Using address 10.12.0.10:6379 set in the environment variable RAY_ADDRESS
2025-01-06 12:44:43,542	INFO worker.py:1634 -- Connecting to existing Ray cluster at address: 10.12.0.10:6379...
2025-01-06 12:44:43,551	INFO worker.py:1810 -- Connected to Ray cluster. View the dashboard at 10.12.0.10:8265 
[0, 1, 4, 9]
2025-01-06 12:44:47,830	SUCC cli.py:63 -- ------------------------------------------
2025-01-06 12:44:47,830	SUCC cli.py:64 -- Job 'raysubmit_fuBdjGnecFggejhR' succeeded
2025-01-06 12:44:47,830	SUCC cli.py:65 -- ------------------------------------------
```

> ### Create a Ray Cluster

Reviewer: Might make sense to show the "Create a Ray cluster" example before "ray job submit" examples, since job submission (without yaml) relies on the cluster creation

Author: Done!
